### PR TITLE
WIP: Enable and Use facets & fields for search

### PR DESCRIPTION
**Pull request by plutonik-a**

### CODE REVIEW

File: modules/search.xqm [hunk 1]
```diff
@@ -47,13 +47,13 @@ declare variable $search:SECTIONS := map {
         map {
             "id": "visits",
             "query": function($query as xs:string) {
-                collection($config:VISITS_COL)//visits[ft:query(., $query, $search:ft-query-options)]
+                collection($config:VISITS_COL)//visits[ft:query(., 'content:' || $query, $search:ft-query-options)]
```
Comment (joewiz) [inline modules/search.xqm:49]: To account for keyword searches containing more than one word, the query string should be wrapped in parentheses. E.g., from hsg-facets: `'hsg-fulltext:(' || $keywords || ')'`.

**Comment by tuurma** [comment]
@joewiz I've tried to refactor along the lines you've sketched in the prototype. Quite advanced there but would need to thoroughly review which collections the search needs to run through and what elements the indexes to set on. Since we don't have much time, I'd avoid changing the default indexed "units", e.g. your prototype for `other-publications/secretary-bios` sets it to `body` while it has been `div` previously which has consequences for display. On the other hand if the whole larger collection (like other-publications) could use the same indexed "unit" we could streamline the search, so I agree fully it should be done in longer perspective.

Where we currently stand, I can say that:
* we don't need memsort anymore
* all searches are executed on Lucene fields, not the default text index
* plugging in new collections into the search is very easy (e.g. administrative-timeline) and only requires that it has the set of required Lucene fields defined

I will test some more today and prepare a table for collections, categories, publications and basic index "unit" for all these for review if that's okay? Already have a stub of it but need to go over once more, checking the index definitions.

**Comment by joewiz** [comment]
@tuurma Thank you for this update!

> we don't need memsort anymore

That's great news. That was one of our biggest motivations for commissioning the fields facility. We'll really benefit from the improved reliability and speed and lower maintenance and processor overhead of fields.

> all searches are executed on Lucene fields, not the default text index

I see branches with xconf changes in the frus, pocom, travels, and visits repos, but I haven't seen branches for the other repositories that have full text indexes. 

> plugging in new collections into the search is very easy (e.g. administrative-timeline) and only requires that it has the set of required Lucene fields defined

The hsg-facets xar has a reference xconf file in `data/administrative-timeline/timeline/collection.xconf` defining the Lucene fields, as well as a query on the administrative-timeline TEI in `search.xq` line 204. 

In other areas:

What is the status of migrating the "filter by section" section of the hsg search page to use eXist's facets facility? This is an important part of the "Upgrade the current implementation of the HSG search page to use eXist 5’s facets and fields feature" projec and was demonstrated in the hsg-facets reference project.

I see caching has been commented out. Disabling caching makes sense for development purposes and for testing this PR. Should we test this PR first once it's ready and approved, reenable caching for a 2nd round of tests and approvals, then merge and go live?

**Comment by tuurma** [comment]
> I see branches with xconf changes in the frus, pocom, travels, and visits repos, but I haven't seen branches for the other repositories that have full text indexes.

Branches for other repos will follow shortly

> What is the status of migrating the "filter by section" section of the hsg search page to use eXist's facets facility? This is an important part of the "Upgrade the current implementation of the HSG search page to use eXist 5’s facets and fields feature" project and was demonstrated in the hsg-facets reference project.

These already use the hsg-section and hsg-publication facets.

> I see caching has been commented out. Disabling caching makes sense for development purposes and for testing this PR. Should we test this PR first once it's ready and approved, reenable caching for a 2nd round of tests and approvals, then merge and go live?

Indeed, temporarily disabled, will reenable it before removing WIP mark from this PR.

**Comment by tuurma** [comment]
@joewiz 

As I'm going through indexes for each publication, I note some differences in results returned

Searching within `countries`, e.g. https://history.state.gov/search?q=Indochina&within=countries&sort-by=date-asc returns 8 hits but in two articles, for Cambodia and Vietnam. Old indexes were defined on tei:div but hsg-facets defines it on tei:body. With this approach we'll only have 2 hits instead of 8. Please confirm this is the desired outcome?

**Comment by tuurma** [comment]
In *conferences* previously other types of divisions would be returned by search, e.g. subchapters cf. https://history.state.gov/search?q=china&within=conferences&sort-by=date-asc and hit for https://history.state.gov/conferences/2006-china-cold-war/brzezinski 

This kind of matches would not be found by current version which indexes only "section" divs. Searching for *china* as above thus presents 14 hits instead of 35.

**Comment by joewiz** [comment]
@tuurma Sounds good!

I can confirm that the new results in `countries` due to the switch from `tei:div` to `tei:body` is preferred, since this publication displays one article per page. 

As to `conferences`, you're right that divs in this collection have numerous `@type` values besides `section`, so this value would skip over divs with important content. Since the collection is not particularly standardized in its use of the `@type` attribute, I would suggest `@xml:id and not(tei:div/@xml:id)` as the criterion for inclusion in the `hsg-fulltext` field.

**Comment by tuurma** [comment]
In `frus-history` the hits for div[@type="part"] previously present are now excluded due to the new index configuration.

**Comment by tuurma** [comment]
In `hac` hits for div[@type="chapter"] previously present are now excluded due to the new index configuration, which seems perfectly right, given these were duplicating the nested "subchapter" hits.

e.g. https://history.state.gov/search?q=myanmar&within=about&sort-by=date-asc

**Comment by tuurma** [comment]
In `faq` as far as I can see division type is set to *section* not *question* as suggested by hsg-facets, so I'll change accordingly

**Comment by tuurma** [comment]
Since other-publications/secretary-bios are now indexed on body, some multiple matches are now compressed into one, e.g. https://history.state.gov/search?start=4&q=japan&within=department&sort-by=date-asc

image

vs

image

**Comment by joewiz** [comment]
@windauer As promised, I've prepared a PR https://github.com/HistoryAtState/hsg-project/pull/55 to remove the memsort module from hsg-project in preparation for facets & fields.

**Comment by joewiz** [comment]
Superseded by #418.

**Pull request closed without merging**